src/Microsoft.Extensions.Primitives/StringSegment.cs

```diff
@@ -633,5 +584,42 @@ public override string ToString()
         {
             return Value ?? string.Empty;
         }
+
+        // Methods that do no return (i.e. throw) are not inlined
+        // https://github.com/dotnet/coreclr/pull/6103
+        private static void ThrowInvalidArguments(string buffer, int offset, int length)
```

Comment:
Need to split into a `GetXxx` method with the `if` branching that creates the exception and a `ThrowXxx` method that just throws the result of the Get.

The Non-returning optimization (which also doesn't push and pop stack); also counts the number of blocks in the method and doesn't kick in if there is more than one; which the `if`s will cause.

Reply:
On doing this PR in the first iteration I just tried to optimize the generated asm, then tweaked based on benchmark-numbers.

Here the JIT produces the same code as with the GetXxx and ThrowXxx methods.
Good to know the hint with the block-counting on non-returning methods -- thanks!
But it seems that isn't the case here...

For reference:

```c#
public class Program
{
    public static int Main(string[] args)
    {
        var res = Do("abc", 1, 2);

        return res != null ? 0 : 1;
    }

    [MethodImpl(MethodImplOptions.NoInlining)]
    private static StringSegment Do(string buffer, int offset, int length)
    {
        return new StringSegment(buffer, offset, length);
    }
}
```
With ThrowXxx + GetXxx (baseline):
```asm
; Assembly listing for method ConsoleApplication.Program:Do(ref,int,int):struct
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rbp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T02] (  7,  5   )     ref  ->  rdi         class-hnd
;  V01 arg1         [V01,T01] (  8,  5   )     int  ->  rsi        
;  V02 arg2         [V02,T03] (  6,  4   )     int  ->  rdx        
;  V03 tmp0         [V03,T00] (  5, 10   )  struct (16) [rbp-0x10]   do-not-enreg[SFBR] multireg-ret must-init
;# V04 OutArgs      [V04    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
;  V05 cse0         [V05,T04] (  6,  3   )     int  ->  rax        
;
; Lcl frame size = 16

G_M15030_IG01:
       55                   push     rbp
       4883EC10             sub      rsp, 16
       C5F877               vzeroupper 
       488D6C2410           lea      rbp, [rsp+10H]
       33C0                 xor      rax, rax
       488945F0             mov      qword ptr [rbp-10H], rax

G_M15030_IG02:
       488D45F0             lea      rax, bword ptr [rbp-10H]

G_M15030_IG03:
       C4E17957C0           vxorpd   xmm0, xmm0
       C4E17A7F00           vmovdqu  qword ptr [rax], xmm0

G_M15030_IG04:
       4885FF               test     rdi, rdi
       7425                 je       SHORT G_M15030_IG07
       8B4708               mov      eax, dword ptr [rdi+8]
       3BC6                 cmp      eax, esi
       721E                 jb       SHORT G_M15030_IG07
       2BC6                 sub      eax, esi
       3BC2                 cmp      eax, edx
       7218                 jb       SHORT G_M15030_IG07

G_M15030_IG05:
       48897DF0             mov      gword ptr [rbp-10H], rdi
       8975F8               mov      dword ptr [rbp-08H], esi
       8955FC               mov      dword ptr [rbp-04H], edx
       488B45F0             mov      rax, gword ptr [rbp-10H]
       488B55F8             mov      rdx, qword ptr [rbp-08H]

G_M15030_IG06:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret      

G_M15030_IG07:
       E820FBFFFF           call     Microsoft.Extensions.Primitives.StringSegment:ThrowInvalidArguments(ref,int,int)
       CC                   int3     

; Total bytes of code 81, prolog size 19 for method ConsoleApplication.Program:Do(ref,int,int):struct
; ============================================================
```

Just with ThrowXxx (this PR):
```asm
; Assembly listing for method ConsoleApplication.Program:Do(ref,int,int):struct
; Emitting BLENDED_CODE for X64 CPU with AVX
; optimized code
; rbp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 arg0         [V00,T02] (  7,  5   )     ref  ->  rdi         class-hnd
;  V01 arg1         [V01,T01] (  8,  5   )     int  ->  rsi        
;  V02 arg2         [V02,T03] (  6,  4   )     int  ->  rdx        
;  V03 tmp0         [V03,T00] (  5, 10   )  struct (16) [rbp-0x10]   do-not-enreg[SFBR] multireg-ret must-init
;# V04 OutArgs      [V04    ] (  1,  1   )  lclBlk ( 0) [rsp+0x00]  
;  V05 cse0         [V05,T04] (  6,  3   )     int  ->  rax        
;
; Lcl frame size = 16

G_M15029_IG01:
       55                   push     rbp
       4883EC10             sub      rsp, 16
       C5F877               vzeroupper 
       488D6C2410           lea      rbp, [rsp+10H]
       33C0                 xor      rax, rax
       488945F0             mov      qword ptr [rbp-10H], rax

G_M15029_IG02:
       488D45F0             lea      rax, bword ptr [rbp-10H]

G_M15029_IG03:
       C4E17957C0           vxorpd   xmm0, xmm0
       C4E17A7F00           vmovdqu  qword ptr [rax], xmm0

G_M15029_IG04:
       4885FF               test     rdi, rdi
       7425                 je       SHORT G_M15029_IG07
       8B4708               mov      eax, dword ptr [rdi+8]
       3BC6                 cmp      eax, esi
       721E                 jb       SHORT G_M15029_IG07
       2BC6                 sub      eax, esi
       3BC2                 cmp      eax, edx
       7218                 jb       SHORT G_M15029_IG07

G_M15029_IG05:
       48897DF0             mov      gword ptr [rbp-10H], rdi
       8975F8               mov      dword ptr [rbp-08H], esi
       8955FC               mov      dword ptr [rbp-04H], edx
       488B45F0             mov      rax, gword ptr [rbp-10H]
       488B55F8             mov      rdx, qword ptr [rbp-08H]

G_M15029_IG06:
       488D6500             lea      rsp, [rbp]
       5D                   pop      rbp
       C3                   ret      

G_M15029_IG07:
       E8F0FAFFFF           call     Microsoft.Extensions.Primitives.StringSegment:ThrowInvalidArguments(ref,int,int)
       CC                   int3     

; Total bytes of code 81, prolog size 19 for method ConsoleApplication.Program:Do(ref,int,int):struct
; ============================================================
```

---

A difference in the dasm is shown when the exception-path is taken, with the extra method `GetInvalidArgumentException` that is JITed. But this isn't relevant to perf.

Reply:
Nice, maybe it has improved.

Another flag to switch on when generating Jit output is
```
set COMPlus_JitPrintInlinedMethods=1
```
And it will give you the reasons why or why not the Jit chose to inline. For exception throws you want it to be saying `does not return` as the reason

Reply:
Cool 😄 So it's much better than reading through the whole JIT-dump. Thank you!

Reply:
This PR shows `[FAILED: unprofitable inline] Microsoft.Extensions.Primitives.StringSegment:ThrowInvalidArguments(ref,int,int)`

With the ThrowXxx + GetXxx it shows `[FAILED: does not return] Microsoft.Extensions.Primitives.StringSegment:ThrowInvalidArguments(ref,int,int)`

So it is coincidentally the same asm, and you were right with the block-counting taken into account for that.

Although the variant with the GetXxx-method seems to be more correct (in the sense of no-return) I would not mind this in that case, because
* the JIT produces the same code
* the C# code is more readable with just one ThrowXxx-method (at least for me)

Anyway, I also wouldn't mind to change it, if you would prefere this.

Reply:
Here's an interesting break down on the whys https://github.com/dotnet/coreclr/issues/17024#issuecomment-374015746

Reply:
Cool 😄  Thanks!

Reply:
Updated the PR to use ThrowXxx + GetXxx methods. It's the correct approach to handle this.

Is a local function for the GetXxx ok? Or should I prefer a regular method?
That the args are passed via stack and not via registers shouldn't matter on the exceptional-path.

Reply:
Local function is fine; as the GetXxx method won't be in stack trace.

Local functions print very strangely in stack traces :)